include/ppx/knob.h

```diff
@@ -0,0 +1,166 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/log.h"
+
+#include <functional>
+#include <iostream>
+#include <map>
+#include <unordered_map>
+#include <vector>
+
+// ---------------------------------------------------------------------------------------------
+// KnobType
+// ---------------------------------------------------------------------------------------------
+
+enum class KnobType
+{
+    Unknown,
+    Bool_Checkbox
+};
+
+std::ostream& operator<<(std::ostream& strm, KnobType kt);
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// KnobConfigs
+// ---------------------------------------------------------------------------------------------
+
+struct KnobConfig
```

Comment:
Why use a struct for that?
If you add a multi-select knob, will this struct be expanded to add more types?

What about a create function with forwarding, so you have:

```cpp
KnobBoolean(const std::string& name, const std::string& description, bool defaultValue);
KnobMultipleChoice(const std::string& name, const std::string& description, const std::vector<std::string>& choices);

template<typename T, typename... Args>
KnobManager::createKnob(Knob *parent, Args... args) {
    T *knob = new T(std::forward(...args));
    if (parent != nullptr) {
        parent.addChild(knob);
    } else {
        mRoots.push_back(knob); // The order here will be your draw order
    }
    return knob;
}
```

Also, if you follow imgui style, `KnobManager` could be `KnobLayout` or `KnobGroup`,
and order of creation would dictate the order:

```cpp
KnobGroup group;
group.createKnob<bool>(nullptr, "firstNode", ...)
Knob* parent = group.createKnob<bool>(nullptr, "secondNode", ...)
group.createKnob(parent, "child1", ...);
group.createKnob(parent, "child2", ...);
```

We could also argue the `KnobGroup` could be a knob, and you'd have the templated `createKnob` being part of Knob, yielding to the following usage, which might be cleaner:

```cpp
KnobGroup group;
group.createKnob<bool>(nullptr, "firstNode", ...)
Knob* parent = group.createKnob<bool>(nullptr, "secondNode", ...)
parent.createKnob(parent, "child1", ...);
parent.createKnob(parent, "child2", ...);

group.draw();
```

Reply:
Well, I do have a separate branch https://github.com/hysw/bigwheels/tree/tianc/knob-template-shenanigans for almost all these comments, just that I don't want to make a competing pull request

I think @angela28chen is still learning more C++ stuff, so template and virtual function might be less familiar.